include/swift/AST/AccessScopeChecker.h

```diff
@@ -23,47 +23,79 @@
 
 namespace swift {
 
-class AbstractStorageDecl;
-class ExtensionDecl;
+class BoundGenericType;
+class ComponentIdentTypeRepr;
+class NominalType;
 class SourceFile;
-class ValueDecl;
+class TypeAliasType;
 
 class AccessScopeChecker {
   const SourceFile *File;
   bool TreatUsableFromInlineAsPublic;
 
-protected:
-  ASTContext &Context;
   Optional<AccessScope> Scope = AccessScope::getPublic();
 
   AccessScopeChecker(const DeclContext *useDC,
                      bool treatUsableFromInlineAsPublic);
-  bool visitDecl(ValueDecl *VD);
-};
-
-class TypeReprAccessScopeChecker : private ASTWalker, AccessScopeChecker {
-  TypeReprAccessScopeChecker(const DeclContext *useDC,
-                             bool treatUsableFromInlineAsPublic);
-
-  bool walkToTypeReprPre(TypeRepr *TR) override;
-  bool walkToTypeReprPost(TypeRepr *TR) override;
+  bool visitDecl(const ValueDecl *VD);
 
 public:
   static Optional<AccessScope>
   getAccessScope(TypeRepr *TR, const DeclContext *useDC,
                  bool treatUsableFromInlineAsPublic = false);
+  static Optional<AccessScope>
+  getAccessScope(Type T, const DeclContext *useDC,
+                 bool treatUsableFromInlineAsPublic = false);
 };
 
-class TypeAccessScopeChecker : private TypeWalker, AccessScopeChecker {
-  TypeAccessScopeChecker(const DeclContext *useDC,
-                         bool treatUsableFromInlineAsPublic);
+/// Walks a Type to find all NominalTypes, BoundGenericTypes, and
+/// TypeAliasTypes.
+class TypeDeclFinder : public TypeWalker {
```

Comment:
Maybe these should live in their own header file?

Reply:
Yeah, if anything the original AccessScopeChecker should just sink into TypeCheckAccess.cpp. It's not reusable after all.

Reply:
I'll come back and fix this in a separate PR, since it doesn't strictly need to go into 5.1.